src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -352,21 +352,29 @@ internal static class ConstraintsHelper
             this Symbol containingSymbol,
             Binder binder,
             ImmutableArray<TypeParameterSymbol> typeParameters,
+            TypeParameterListSyntax typeParameterList,
             SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses,
             Location location,
             DiagnosticBag diagnostics)
         {
-            if (typeParameters.Length == 0 || constraintClauses.Count == 0)
+            if (typeParameters.Length == 0)
             {
                 return ImmutableArray<TypeParameterConstraintClause>.Empty;
             }
 
+            if (constraintClauses.Count == 0)
+            {
+                ImmutableArray<TypeParameterConstraintClause> defaultClauses = binder.GetDefaultTypeParameterConstraintClauses(typeParameterList);
+
+                return defaultClauses.IsEmpty() ? ImmutableArray<TypeParameterConstraintClause>.Empty : defaultClauses;
```

Comment:
>defaultClauses.IsEmpty() ? ImmutableArray<TypeParameterConstraintClause>.Empty : defaultClauses [](start = 23, length = 95)

I don't understand why this isn't just `defaultClauses`.  Are you trying to implement an optimization that you assume `ArrayBuilder` doesn't perform?

Reply:
> I don't understand why this isn't just defaultClauses.  Are you trying to implement an optimization that you assume ArrayBuilder doesn't perform?

```defaultClauses``` here contains at least one item. If all items do not contain any significant data, we replace it with an empty array so that later phases don't have to worry about doing more analysis. This mimics what we would get prior to this PR.




---
In reply to: [273178358](https://github.com/dotnet/roslyn/pull/34797#discussion_r273178358) [](ancestors = 273178358)

Reply:
How can "`defaultClauses` here contains at least one item" when `defaultClauses.IsEmpty()`?

Reply:
> How can "defaultClauses here contains at least one item" when defaultClauses.IsEmpty()?

This is an extension method in ```TypeParameterConstraintClauseExtensions``` type, not a property on an ImmutableArray that checks it's length.

---
In reply to: [273197979](https://github.com/dotnet/roslyn/pull/34797#discussion_r273197979) [](ancestors = 273197979)